src/Store.ts

```diff
@@ -0,0 +1,581 @@
+	//		objects (this can improve performance by avoiding prototype setting),
+	Model: NewableStoreModel;
+
+	// parse: Function
+	//		One can provide a parsing function that will permit the parsing of the data. By
+	//		default we assume the provide data is a simple JavaScript array that requires
+	//		no parsing (subclass stores may provide their own default parse function)
+	parse: (...args: any[]) => any;
+
+	// summary:
+	//		Stores an object
+	// object: Object
+	//		The object to store.
+	// directives: dstore/Store.PutDirectives?
+	//		Additional directives for storing objects.
+	// returns: Object
+	//		The object that was stored, with any changes that were made by
+	//		the storage system (like generated id)
+	put: (object: any, directives?: dstore.PutDirectives) => any;
+
+	// queryAccessors: Boolean
+	//		Indicates if client-side query engine filtering should (if the store property is true)
+	//		access object properties through the get() function (enabling querying by
+	//		computed properties), or if it should (by setting this to false) use direct/raw
+	// 		property access (which may more closely follow database querying style).
+	queryAccessors: boolean =  true;
+
+	// summary:
+	//		Deletes an object by its identity
+	// id: Number
+	//		The identity to use to delete the object
+	remove: (id: any) => Promise<Object>;
+
+	// queryLog: __QueryLogEntry[]
+	//		The query operations represented by this collection
+	queryLog: dstore.QueryLogEntry<any> [];
+
+	storage: Evented;
+
+	// stringify: Function
+	//		For stores that serialize data (to send to a server, for example) the stringify
+	//		function can be specified to control how objects are serialized to strings
+	stringify: () => any;
+
+	constructor(options?: StoreArgs) {
+		super();
+		this.autoEmitEvents = true;
+		this.autoEmitHandles = [];
+		this.idProperty = 'id';
+		this.queryLog = [];
+		this.Filter = Filter;
+		// Set defaults
+		// perform the mixin
+		options && lang.mixin(this, options);
+		// TODO - Don't do this, it's a workaround to make inheritance from this class via
+		// declare work properly.
+		for (let name in this) {
+			let value = this[name];
+			if (Object.prototype.toString.call(value) === '[object Function]') {
+				value.nom = name;
+			}
+		}
+
+		// TODO - This is designed specifically for working with declare-based inheritance and needs to be reworked.
+		if (this.Model && this.Model.createSubclass) {
+			// we need a distinct model for each store, so we can
+			// save the reference back to this store on it.
+			// we always create a new model to be safe.
+			this.Model = this.Model.createSubclass([]).extend({
+				// give a reference back to the store for saving, etc.
+				_store: this
+			});
+		}
+
+		// the object the store can use for holding any local data or events
+		this.storage = new Evented();
+		const store = this;
+		if (this.autoEmitEvents) {
+			// emit events when modification operations are called
```

Comment:
IIRC there's a TODO about these handles in another module already, but should we add one here too?